### PR TITLE
planner: refine collation handling for between

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Refine collation handling for between. The previous way is very hack. 

### What is changed and how it works?
Convert the argument collation for between, then we can rewrite it to `equal`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -1686,19 +1686,19 @@ func (er *expressionRewriter) betweenToExpression(v *ast.BetweenExpr) {
 		return
 	}
 
+	expr = expression.BuildCastCollationFunction(er.sctx, expr, coll)
```
Comment (winoros) [inline planner/core/expression_rewriter.go:1689]: Do we need to align the `Coercibility`?

Comment (wjhuang2016) [reply planner/core/expression_rewriter.go:1689]: No

**Review by winoros — APPROVED** [review]

**Review by xiongjiwei — APPROVED** [review]

**Comment by wjhuang2016** [comment]
@ti-chi-bot

**Pull request merged**